service/history/historyReplicator.go

```diff
@@ -68,167 +77,238 @@ func newHistoryReplicator(shard ShardContext, historyEngine *historyEngineImpl,
+		// no matter what error happen, we need to retry
+		release(err)
+		return err
+	}
+	currentLastWriteVersion := currentMsBuilder.GetLastWriteVersion()
+	currentRunID := currentMsBuilder.GetExecutionInfo().RunID
+	release(nil)
+
+	// we can also use the start version
+	if currentLastWriteVersion > incomingVersion {
+		logger.Infof("Dropping replication task. Current RunID: %v, Current LastWriteVersion: %v, Incoming Version: %v.",
+			currentRunID, currentLastWriteVersion, incomingVersion)
+		return nil
+	}
+	// currentLastWriteVersion <= incomingVersion
+	logger.Debugf("Retrying replication task. Current RunID: %v, Current LastWriteVersion: %v, Incoming Version: %v.",
+		currentRunID, currentLastWriteVersion, incomingVersion)
+	return ErrRetryEntityNotExists
+}
 
-		// Check for duplicate processing of replication task
-		if firstEvent.GetEventId() < msBuilder.GetNextEventID() {
-			replicationState := msBuilder.GetReplicationState()
-			r.metricsClient.IncCounter(metrics.ReplicateHistoryEventsScope, metrics.StaleReplicationEventsCounter)
-			logger.Debugf("Dropping replication task.  State: {NextEvent: %v, Version: %v, LastWriteV: %v, LastWriteEvent: %v}",
-				msBuilder.GetNextEventID(), replicationState.CurrentVersion,
-				replicationState.LastWriteVersion, replicationState.LastWriteEventID)
-			return nil
+func (r *historyReplicator) ApplyOtherEventsVersionChecking(context *workflowExecutionContext, msBuilder mutableState,
+	request *h.ReplicateEventsRequest, logger bark.Logger) (mutableState, error) {
+	var err error
+	// check if to buffer / drop / conflict resolution
+	incomingVersion := request.GetVersion()
+	replicationInfo := request.ReplicationInfo
+	rState := msBuilder.GetReplicationState()
+	if rState.LastWriteVersion > incomingVersion {
+		// Replication state is already on a higher version, we can drop this event
+		// TODO: We need to replay external events like signal to the new version
+		logger.Warnf("Dropping stale replication task. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		return nil, nil
+	} else if rState.LastWriteVersion < incomingVersion {
+		// Check if this is the first event after failover
+		logger.Infof("First Event after replication. CurrentV: %v, LastWriteV: %v, LastWriteEvent: %v, IncomingV: %v.",
+			rState.CurrentVersion, rState.LastWriteVersion, rState.LastWriteEventID, incomingVersion)
+		previousActiveCluster := r.clusterMetadata.ClusterNameForFailoverVersion(rState.LastWriteVersion)
+		ri, ok := replicationInfo[previousActiveCluster]
+		if !ok {
+			logger.Errorf("No ReplicationInfo Found For Previous Active Cluster. Previous Active Cluster: %v, Request Source Cluster: %v, Request ReplicationInfo: %v.",
+				previousActiveCluster, request.GetSourceCluster(), request.ReplicationInfo)
+			// TODO: Handle missing replication information, #840
+			return nil, ErrMissingReplicationInfo
```

Comment:
Can you also include the existing comment for this part of the code.

Reply:
copy pasted the existing one: "// Returning BadRequestError to force the message to land into DLQ"
is this the comment you are talking about?